src/nnet3/nnet-optimize.cc

```diff
@@ -432,6 +433,134 @@ void CachingOptimizingCompiler::UpdateCache(const ComputationRequest *request,
+                              == allocate_from_other);
+  ExpectToken(is, binary, "<MinDerivTime>");
+  int32 min_deriv_time;
+  ReadBasicType(is, binary, &min_deriv_time);
+  read_cache = read_cache && (opt_config_.min_deriv_time
+                              == min_deriv_time);
+  ExpectToken(is, binary, "<MaxDerivTime>");
+  int32 max_deriv_time;
+  ReadBasicType(is, binary, &max_deriv_time);
+  read_cache = read_cache && (opt_config_.max_deriv_time
+                              == max_deriv_time);
+  ExpectToken(is, binary, "</NnetOptimizeOptions>");
+  
+  if (read_cache) {
+    size_t computation_cache_size;
+    ExpectToken(is, binary, "<ComputationCacheSize>");
+    ReadBasicType(is, binary, &computation_cache_size);
+    KALDI_ASSERT(computation_cache_size >= 0);
+    computation_cache_.clear();
+    access_queue_.clear();
+    ExpectToken(is, binary, "<ComputationCache>");
+    for (size_t c = 0; c < computation_cache_size; c++) {
+      ComputationRequest *request = new ComputationRequest();
+      request->Read(is, binary);
+      NnetComputation *computation = new NnetComputation();
+      computation->Read(is, binary);
+      UpdateCache(request, computation);
+    }
+  }
+}
+
+void CachingOptimizingCompiler::WriteCache(std::ostream &os, bool binary) const {
+  WriteToken(os, binary, "<NnetOptimizeOptions>");
+  WriteToken(os, binary, "<Optimize>");
+  WriteBasicType(os, binary, opt_config_.optimize);
+  WriteToken(os, binary, "<Consolidate_model_update>");
+  WriteBasicType(os, binary, opt_config_.consolidate_model_update);
+  WriteToken(os, binary, "<PropagateInPlace>");
+  WriteBasicType(os, binary, opt_config_.propagate_in_place);
+  WriteToken(os, binary, "<BackpropInPlace>");
+  WriteBasicType(os, binary, opt_config_.backprop_in_place);
+  WriteToken(os, binary, "<ConvertAddition>");
+  WriteBasicType(os, binary, opt_config_.convert_addition);
+  WriteToken(os, binary, "<RemoveAssignments>");
+  WriteBasicType(os, binary, opt_config_.remove_assignments);
+  WriteToken(os, binary, "<AllowLeftMerge>");
+  WriteBasicType(os, binary, opt_config_.allow_left_merge);
+  WriteToken(os, binary, "<AllowRightMerge>");
+  WriteBasicType(os, binary, opt_config_.allow_right_merge);
+  WriteToken(os, binary, "<InitializeUndefined>");
+  WriteBasicType(os, binary, opt_config_.initialize_undefined);
+  WriteToken(os, binary, "<MoveSizingCommands>");
+  WriteBasicType(os, binary, opt_config_.move_sizing_commands);
+  WriteToken(os, binary, "<AllocateFromOther>");
+  WriteBasicType(os, binary, opt_config_.allocate_from_other);
+  WriteToken(os, binary, "<MinDerivTime>");
+  WriteBasicType(os, binary, opt_config_.min_deriv_time);
+  WriteToken(os, binary, "<MaxDerivTime>");
+  WriteBasicType(os, binary, opt_config_.max_deriv_time);
+  WriteToken(os, binary, "</NnetOptimizeOptions>");
+
+  WriteToken(os, binary, "<ComputationCacheSize>");
+  WriteBasicType(os, binary, computation_cache_.size());
```

Comment:
It's dangerous to write types of machine-dependent type directly to disk like this.  Use a temporary int32.